vault/resource_kmip_secret_backend.go

```diff
@@ -0,0 +1,222 @@
+	path := d.Get("path").(string)
+	description := d.Get("description").(string)
+	defaultTLSClientTTL := d.Get("default_tls_client_ttl").(int)
+
+	log.Printf("[DEBUG] Mounting KMIP backend at %q", path)
+	if err := client.Sys().Mount(path, &api.MountInput{
+		Type:        "kmip",
+		Description: description,
+		Config: api.MountConfigInput{
+			DefaultLeaseTTL: fmt.Sprintf("%d", defaultTLSClientTTL),
+		},
+	}); err != nil {
+		return fmt.Errorf("error mounting to %q: %s", path, err)
+	}
+
+	log.Printf("[DEBUG] Mounted KMIP backend at %q", path)
+	d.SetId(path)
+
+	return kmipSecretBackendUpdate(d, meta)
+}
+
+func kmipSecretBackendUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+	tune := api.MountConfigInput{}
+
+	if d.HasChange("default_tls_client_ttl") || d.HasChange("description") {
+		tune.DefaultLeaseTTL = fmt.Sprintf("%ds", d.Get("default_tls_client_ttl"))
+		description := d.Get("description").(string)
+		tune.Description = &description
+
+		log.Printf("[DEBUG] Updating mount for %q", path)
+		err := client.Sys().TuneMount(path, tune)
+		if err != nil {
+			return fmt.Errorf("error updating mount for %q, err=%w", path, err)
+		}
+		log.Printf("[DEBUG] Updated mount for %q", path)
+	}
+
+	data := map[string]interface{}{}
+	configPath := fmt.Sprintf("%s/config", path)
+	log.Printf("[DEBUG] Updating %q", configPath)
+
+	for _, k := range kmipAPIFields {
+		if d.HasChange(k) {
+			if v, ok := d.GetOk(k); ok {
+				switch v.(type) {
+				case *schema.Set:
+					data[k] = util.TerraformSetToStringArray(v)
+				default:
+					data[k] = v
+				}
+			}
+		}
+	}
+
+	if _, err := client.Logical().Write(configPath, data); err != nil {
+		return fmt.Errorf("error updating KMIP config %q: %s", configPath, err)
+	}
+
+	log.Printf("[DEBUG] Updated %q", configPath)
+
+	return kmipSecretBackendRead(d, meta)
+}
+
+func kmipSecretBackendRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	path := d.Id()
+	log.Printf("[DEBUG] Reading KMIP config at %s/config", path)
+	resp, err := client.Logical().Read(path + "/config")
+	if err != nil {
+		return fmt.Errorf("error reading KMIP config at %q/config: %s", path, err)
+	}
+
+	if resp == nil {
+		log.Printf("[WARN] KMIP config not found, removing from state")
+		d.SetId("")
+
+		return nil
+	}
+
+	for _, k := range kmipAPIFields {
+		if err := d.Set(k, resp.Data[k]); err != nil {
+			return fmt.Errorf("error setting state key %q on KMIP config, err=%w", k, err)
+		}
+	}
+
+	return nil
+}
+
+func kmipSecretBackendDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+	log.Printf("[DEBUG] Unmounting KMIP backend %q", path)
+
+	if err := client.Sys().Unmount(path); err != nil {
+		if util.Is404(err) {
+			log.Printf("[WARN] %q not found, removing from state", path)
+			d.SetId("")
+
+			return fmt.Errorf("error unmounting KMIP backend from %q: %s", path, err)
```

Comment:
We should always use the `%w` verb whenever the error supports it.

Reply:
Fixed in 3686e05